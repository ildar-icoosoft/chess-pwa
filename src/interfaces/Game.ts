import GameStatus from "../types/GameStatus";
import User from "./User";
import { PieceColor } from "../types/PieceColor";

export default interface Game {
  id: number;
  aiLevel: number;
  clockLimit: number;
  clockIncrement: number;
  initialFen: string;
  turn: PieceColor;
  wtime: number;
  btime: number;
  moves: string;
  status: GameStatus;
  white: User | null;
  black: User | null;
  winner: PieceColor | null;
}
