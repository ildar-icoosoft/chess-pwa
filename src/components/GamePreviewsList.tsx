/**
 * Renders a chess board using React
 */
import React, { FC } from "react";
import { Board } from "ii-react-chessboard";
import Game from "../interfaces/Game";

const initialBoardFen =
  "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1";

export interface GamePreviewsListProps {
  games?: Game[];
}

export const GamePreviewsList: FC<GamePreviewsListProps> = ({ games = [] }) => {
  return (
    <div>
      {games.map((item) => {
        let fen: string = item.initialFen;
        if (fen === "startpos") {
          fen = initialBoardFen;
        }

        return <Board key={item.id} position={fen} viewOnly={false} />;
      })}
    </div>
  );
};
