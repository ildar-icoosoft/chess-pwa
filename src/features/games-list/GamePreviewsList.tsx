import React, { FC } from "react";
import Game from "../../interfaces/Game";
import css from "./GamePreviewsList.module.scss";
import { GamePreviewsListItem } from "./GamePreviewsListItem";
import { ContentLoadingStatus } from "../../components/ContentLoadingStatus";

export interface GamePreviewsListProps {
  games?: Game[];
  isLoading?: boolean;
  error?: string | null;
  emptyContentMessage?: string;
}

export const GamePreviewsList: FC<GamePreviewsListProps> = ({
  games = [],
  isLoading = false,
  error = null,
  emptyContentMessage,
}) => {
  return (
    <div>
      <ContentLoadingStatus
        isLoading={isLoading}
        error={error}
        isEmpty={games.length === 0}
        emptyContentMessage={emptyContentMessage}
      />
      <div className={css.grid}>
        {games.map((item) => (
          <GamePreviewsListItem game={item} key={item.id} />
        ))}
      </div>
    </div>
  );
};
