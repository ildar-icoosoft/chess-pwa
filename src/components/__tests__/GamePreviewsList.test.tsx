import React from "react";
import TestRenderer from "react-test-renderer";
import { Board } from "ii-react-chessboard";
import { GamePreviewsList } from "../GamePreviewsList";
import mountTest from "../../tests/mountTest";
import Game from "../../interfaces/Game";

const games: Game[] = [
  {
    id: 1,
    initialFen: "startpos",
    wtime: 300000,
    btime: 300000,
    moves: "",
    status: "started",
    white: null,
    black: null,
  },
  {
    id: 2,
    initialFen: "rnbqkbnr/8/8/8/8/8/8/RNBQKBNR w KQkq - 0 1",
    wtime: 300000,
    btime: 300000,
    moves: "",
    status: "started",
    white: null,
    black: null,
  },
];

const initialBoardFen =
  "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1";

describe("GamePreviewsList", () => {
  mountTest(GamePreviewsList);

  it("Snapshot", () => {
    const tree = TestRenderer.create(<GamePreviewsList />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe("children components", () => {
    it("contains Board", () => {
      const testRenderer = TestRenderer.create(<GamePreviewsList />);
      const testInstance = testRenderer.root;

      expect(testInstance.findAllByType(Board).length).toBe(0);

      testRenderer.update(<GamePreviewsList games={games} />);
      expect(testInstance.findAllByType(Board).length).toBe(2);
    });
  });

  describe("children components props", () => {
    describe("Board", () => {
      it("position", () => {
        const testRenderer = TestRenderer.create(
          <GamePreviewsList games={games} />
        );
        const testInstance = testRenderer.root;

        const boards = testInstance.findAllByType(Board);

        expect(boards[0].props.position).toBe(initialBoardFen);
        expect(boards[1].props.position).toBe(
          "rnbqkbnr/8/8/8/8/8/8/RNBQKBNR w KQkq - 0 1"
        );
      });

      it("viewOnly", () => {
        const testRenderer = TestRenderer.create(
          <GamePreviewsList games={games} />
        );
        const testInstance = testRenderer.root;

        const boards = testInstance.findAllByType(Board);

        expect(boards.every((item) => item.props.viewOnly === false)).toBe(
          true
        );
      });
    });
  });
});
