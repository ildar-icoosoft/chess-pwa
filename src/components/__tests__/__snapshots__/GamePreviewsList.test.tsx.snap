// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OngoingGames Snapshot 1`] = `
<div
  className="grid"
/>
`;
